relay/src/transport.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use core::{Transport, MuxedTransport};
+use futures::{stream, prelude::*};
+use message::{CircuitRelay, CircuitRelay_Peer, CircuitRelay_Type};
+use multiaddr::Multiaddr;
+use peerstore::{PeerAccess, PeerId, Peerstore};
+use protocol::{self, RelayConfig};
+use rand::{self, Rng};
+use std::{io, iter::FromIterator, ops::Deref, sync::Arc};
+use tokio_io::{AsyncRead, AsyncWrite};
+use utility::{io_err, Peer, RelayAddr};
+
+#[derive(Debug, Clone)]
+pub struct RelayTransport<T, P> {
+    my_id: PeerId,
+    transport: T,
+    peers: P,
+    relays: Arc<Vec<PeerId>>
+}
+
+impl<T, P, S> Transport for RelayTransport<T, P>
+where
+    T: Transport + Clone + 'static,
+    T::Output: AsyncRead + AsyncWrite,
+    P: Deref<Target=S> + Clone + 'static,
+    S: 'static,
+    for<'a> &'a S: Peerstore
+{
+    type Output = protocol::Output<T::Output>;
+    type Listener = Box<Stream<Item=Self::ListenerUpgrade, Error=io::Error>>;
+    type ListenerUpgrade = Box<Future<Item=(Self::Output, Multiaddr), Error=io::Error>>;
+    type Dial = Box<Future<Item=(Self::Output, Multiaddr), Error=io::Error>>;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<(Self::Listener, Multiaddr), (Self, Multiaddr)> {
+        let rly = RelayConfig::new(self.my_id.clone(), self.transport.clone(), self.peers.clone());
+        let (listener, new_addr) = self.transport
+            .clone()
+            .with_upgrade(rly)
```

Comment:
Isn't this the same when [dialing](https://github.com/libp2p/rust-libp2p/pull/175/files#diff-edc4876e99c7bccb40bba2677c99936fR133), i.e. the underlying transport would use multiplex for example? Additional protocol upgrades like secio are applied in [`and_then`](https://github.com/libp2p/rust-libp2p/pull/175/files#diff-edc4876e99c7bccb40bba2677c99936fR154), no?

One advantage of not allowing listening on a `RelayTransport` would be that the output type can be `AsyncRead + AsyncWrite` again, as dialing will always yield a stream, but the asymmetry feels not quite right to me. Currently, the relay example could also be implemented with `RelayTransport` in `run_listener` and it would just work™. Maybe I am missing something?